### DISCUSSION TITLE
Why CustomUIExtension force selection of diagram

### DISCUSSION THREAD

**Discussion by wss29** [eclipse-glsp/glsp#1186] [Q&A]
hi,
   1. I created a custom ui extension, it is a switch, and when I switch it on it invokes startMonitor method, startMonitor dispatch an action to the server every second
![image](https://github.com/eclipse-glsp/glsp/assets/4537809/218be3ef-54dc-439e-9da7-1b76d3af35b0)
![image](https://github.com/eclipse-glsp/glsp/assets/4537809/58bd775a-c393-4a94-9671-367ee78b6e30)
   2. the action handler refreshed my model and called submissionHandler.submitModel to update the client
![image](https://github.com/eclipse-glsp/glsp/assets/4537809/90b0a53a-38a2-49ad-983c-f868d035a71f)
  3. the problem is that the diagram will always be selected, and the SelectionService of Theia will trigger every second I dispatch my action from the custom ui extension
![image](https://github.com/eclipse-glsp/glsp/assets/4537809/e95d6fa1-2d75-49e6-b503-996d6b6bed1e)
  4. because my other extension also uses SelectionService of Theia , if the diagram always be selected every second, the other extension will not work correctly
 In glsp 1.0 I still use this solution, but the selection service does not always trigger. What do I need to do to make it not trigger the selection when  I dispath action in my ui extension?

**Comment by wss29** [comment]
hi,
   Does anyone care about this problem?  If dispatch action from the server to the client causes a selection event fire, this will cause other theia extension problems, because many situations also use selection service in other extensions, in follow situation extension can not work
1. update glsp model every second in the server,
2. int another extension, right-click a custom file and open a dialog by checking the current select file
you can not operate the extension correctly, because your selection always focuses on the glsp client. when you want to open a dialog by checking to right-click file, you can not get the right result, you must stop the interval in the server and can get the correct parameter by selection service in another extension

**Comment by planger** [comment]
Hi @wss29,

so far I didn't encounter this issue, as it is a rather rare use case to update the model periodically from the server without any user interaction, where the focus would be on the diagram anyway. Thus, I guess not many people will face this issue.

Nevertheless, I'm all for supporting that in GLSP, so I'd be grateful if you could debug your scenario and suggest a fix. It feels it'd be best to avoid the selection update on model update. Which code causes the selection update? Or should the `SelectionService.selection` setter check whether the selection actually changed (`selection !== this.currentSelection`) to guard the update?

**Reply by wss29** [reply]
hi,
     as you say it'd be best to avoid the selection update on model update. my server code as follows
![image](https://github.com/eclipse-glsp/glsp/assets/4537809/b2c4b3c5-cfef-427b-8e48-1a2dc49ef2cf)

     const result = await this.submissionHandler.submitModel('external');
     this.actionDispatcher.dispatchAll(result);
these two line codes cause selection updates. these codes invoke glsp internal core code to update model. I am not sure if is there any method to avoid the selection trigger.

current code seems simple and graceful for updating model periodically from the server if it can avoid selection invoke, or I will change the client code to make it complex and hard. 

maybe most people use static use case, so it is rare to update model periodically, but in Industrial system dynamic use case is very common, I am trying to use glsp in those systems, and if glsp is good enough, we will finally change to glsp and drop others

**Reply by planger** [reply]
The two lines you highlighted send an `UpdateModel` action to the client. But it would be good to identify the code on the client that -- as a consequence of receiving the `UpdateModel` action -- set the selection. Can you please check in the call stack of setting `SelectionService.selection` which code is setting the selection? Based on that we can best decide how to avoid this invocation and happily review a PR.
Thank you!

**Reply by wss29** [reply]
hi,
    I debug and get the following invoke stack, you can check if this is what you want, if not, I will debug and get more stack
1. when I open my custom diagram file, I find command-stack.ts file in glsp-client registers a SelectionService at initialize.
![image](https://github.com/eclipse-glsp/glsp/assets/4537809/6abc6a86-7eeb-4ef4-9b80-56261f3c3a40)

2. when I invoke this.submissionHandler.submitModel('external') to update model, it returns two actions. it contains a RequestBound action.
![image](https://github.com/eclipse-glsp/glsp/assets/4537809/88131640-d9b8-4031-a48f-463736b2d532)

3. the RequestBound action is handed by command-stack, and the execute method invokes notifyListeners, which will invoke the selection service register at initialize. the call of selection service makes it always focus on diagram
![image](https://github.com/eclipse-glsp/glsp/assets/4537809/516c1120-2b6c-44de-828e-63761db791ba)
![image](https://github.com/eclipse-glsp/glsp/assets/4537809/312169b1-dc7d-44cb-94fd-4bdf254207d0)

**Reply by planger** [reply]
Thanks for debugging this a bit more! Imho there are two thing that are unclear to me, which we should investigate:
1. Why do we update the selection `onModelRootChanged` to an empty set (`SelectionService.modelRootChanged`)? Imho this should be reconsidered if that's necessary / useful.
2. Why does the update of the selection impact / steel the focus?

I've created [an issue](https://github.com/eclipse-glsp/glsp/issues/1194) to track these two questions. If you happen to have time to look into this further, we appreciate any input or PR! Thanks!
